src/react/utils/useHydrateAtoms.ts

```diff
@@ -7,8 +7,8 @@ type AnyWritableAtom = WritableAtom<unknown, any[], any>
 
 const hydratedMap: WeakMap<Store, WeakSet<AnyWritableAtom>> = new WeakMap()
 
-export function useHydrateAtoms(
-  values: Iterable<readonly [AnyWritableAtom, unknown]>,
+export function useHydrateAtoms<T extends AnyWritableAtom, V = T['read']>(
```

Comment:
It should be something like this. Not sure why it worked before.
```suggestion
export function useHydrateAtoms<T extends AnyWritableAtom, V = T extends WritableAtom<infer Value, any[], any> ? Value : never>(
```

Reply:
Test the above and still not what we need. 

### Code
<img width="801" alt="Screenshot 2023-04-27 at 9 25 13 AM" src="https://user-images.githubusercontent.com/525506/234893045-3600f397-c218-4b01-bf0b-ee997cf0a62a.png">
### Test
<img width="884" alt="Screenshot 2023-04-27 at 9 25 39 AM" src="https://user-images.githubusercontent.com/525506/234893087-71629201-2b2f-4d65-adf8-01dd958b19dc.png">

Need something recursive like this, but this is not working yet either.
```ts
type AnyWritableAtom = WritableAtom<unknown, any[], any>
type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]

type InferAtoms<T extends Iterable<AtomTuple>> = {
  [K in keyof T]: T[K] extends infer A
    ? A extends AnyWritableAtom
      ? AtomTuple<A, ReturnType<A['read']>>
      : [unknown, unknown]
    : [unknown, unknown]
```

### Code
<img width="921" alt="Screenshot 2023-04-27 at 9 26 52 AM" src="https://user-images.githubusercontent.com/525506/234894142-acd65cb0-b2a7-4e75-871a-aa7880ff891c.png">

### Test
<img width="942" alt="Screenshot 2023-04-27 at 9 27 09 AM" src="https://user-images.githubusercontent.com/525506/234894195-bf0fe63a-8b2a-495e-95ed-657540f13dbb.png">

Reply:
I think I have it, but we'd have to use `Array<>` rather than `Iterable<>`
```ts
type InferAtoms<
  T extends Array<AtomTuple>,
  S extends Array<AtomTuple> = []
> = S['length'] extends T['length']
  ? S
  : T extends Array<AtomTuple<infer A>>
  ? A extends AnyWritableAtom
    ? InferAtoms<T, [AtomTuple<A, ReturnType<A['read']>>, ...S]>
    : T
  : T

export function useHydrateAtoms<T extends Array<AtomTuple>>(
  values: InferAtoms<T>,
  options?: Options
) {
```
<img width="945" alt="Screenshot 2023-04-27 at 9 37 10 AM" src="https://user-images.githubusercontent.com/525506/234896905-b89fca9c-1d4c-4627-bd70-4d35770c9c47.png">